### ISSUE TITLE
Add a method to PSBT to compute sighash::Prevouts

### ISSUE THREAD

**Issue by sanket1729** [rust-bitcoin/rust-bitcoin#849]
The most important feature of Psbt's is that they have all the information to compute the sighash. Our new APIs for `taproot_key_spend_signature_hash` and `taproot_script_spend_signature_hash` require a `sighash::Prevouts` struct. This is essentially a collection of all utxos the transaction is spending. 

It would be great to have the following API:

```rust
impl PartialSignedTransaction {
	pub fn to_prevouts(&self) -> sighash::Prevouts {
		/// Always return the  Prevouts::All variant as psbts must have UTXO information
	}
}
```

**Comment by violet360** [comment]
Hey @sanket1729 , new to rust and this project, but this seems very interesting one to start my journey with, looking forward to explore and raise a PR.

**Comment by violet360** [comment]
So I read a bit about psbts, ( [here](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki),  [here](https://www.youtube.com/watch?v=H6xZSRDXUiU&t=570s), and [here](https://github.com/bitcoin/bitcoin/blob/master/doc/psbt.md#psbt-in-general) )started by tracing the class `PartiallySignedTransaction`, I think this is where I will add the function and then tried to map a rough schema of the class
```
PartiallySignedTransaction
    unsigned_tx: Transaction,
    pub version: u32
    pub xpub: BTreeMap<ExtendedPubKey, KeySource>,
    pub proprietary: BTreeMap<raw::ProprietaryKey, Vec<u8>>, // not sure what this is
    pub unknown: BTreeMap<raw::Key, Vec<u8>>, // not sure what this is
    pub inputs: Vec<Input>,
    pub outputs: Vec<Output>,
```

```
Transaction
    pub version: i32,
    pub lock_time: u32,
    pub input: Vec<TxIn>, // TxIn has Utxo info....
    pub output: Vec<TxOut>,
```

```
TxIn (this is what we need)
    pub previous_output: OutPoint,
    pub script_sig: Script,
    pub sequence: u32,
    pub witness: Witness
```

so based on this I wrote some rough code, just tell me am I approaching it right ? if not I will try to think in some other direction.
``` rust
impl PartiallySignedTransaction {
	pub fn to_prevouts(&self) -> sighash::Prevouts {
        	let mut inputs : Vec<TxIn> = self.unsigned_tx.input
        	let mut prevOuts : Vec<TxIn> = Vec::new();
        	for x in 0..inputs.len() {
            		prevOuts.push(inputs[x])
        	}
        	return prevOuts;
    	}
}
```

**Comment by violet360** [comment]
So I read a bit about psbts, ( [here](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki),  [here](https://www.youtube.com/watch?v=H6xZSRDXUiU&t=570s), and [here](https://github.com/bitcoin/bitcoin/blob/master/doc/psbt.md#psbt-in-general) )started by tracing the class `PartiallySignedTransaction`, I think this is where I will add the function and then tried to map a rough schema of the class.
so based on this I wrote some rough snippet, just tell me am I approaching it right ?
``` rust
impl PartiallySignedTransaction {
	pub fn to_prevouts(&self) -> sighash::Prevouts {
        	let mut inputs : Vec<TxIn> = self.unsigned_tx.input;
                return inputs;
    	}
}

**Comment by violet360** [comment]
Thanks for pointing out the cases, went in the wrong direction, here is some revamped code.
```rust

impl PartiallySignedTransaction {
	pub fn to_prevouts(&self) -> sighash::Prevouts {
        let mut prev_outs : Vec<TxOut> = Vec::new(); 
        for i in 0..self.inputs.len() {
            let psbt_input : Input = self.inputs[i];
                if let Some(ref witness_utxo ) = psbt_input.witness_utxo { // check for witness utxos
                    prev_outs.push(&witness_utxo)
                } else if let Some(ref non_witness_utxo) = psbt_input.non_witness_utxo { // if no witness utxos 
                    let vout : usize = psbt.unsigned_tx.input[index].previous_output.vout; // then catch the output index from the psbt input
                    psbt_input(&non_witness_utxo.output[vout]) // and extract and push the corresponding utxo.
                } else {
                    return Err(InputError::MissingUtxo);
                }
            }
            return prev_outs;
    	}
}
```
does it look correct ?

**Comment by violet360** [comment]
Hey, thanks for pointing out the cases, does this have the correct flow ?
```rust
impl PartiallySignedTransaction {
    pub fn to_prevouts(&self) -> sighash::Prevouts {
        let mut prev_outs : Vec<&TxOut> = Vec::new(); 
        for i in 0..self.inputs.len() {
            let psbt_input : Input = self.inputs[i];
                if let Some(ref witness_utxo ) = psbt_input.witness_utxo { // check for witness utxos
                    prev_outs.push(&witness_utxo)
                } else if let Some(ref non_witness_utxo) = psbt_input.non_witness_utxo { // if no witness utxos then catch the non witness part of the input
                    let vout : usize = self.unsigned_tx.input[i].previous_output.vout;
                    prev_outs.push(&non_witness_utxo.output[vout]) // extract and push the corresponding utxo
                }
            }
            return prev_outs;
    	}
}
```

**Comment by tcharding** [comment]
So this is a bit of a curly one because of Rust borrowing rules. I played around with this and was able to return a `Result<Prevouts<&TxOut>, Error>` by first changing the `Prevouts::All` variant to use vec instead of a slice (removes the explicit lifetime also).

```
-    All(&'u [T]),
+    All(Vec<T>),
```
 I was unable to work out _why_ we use a slice currently or what the implications of changing this were? (If you choose to follow this route you could do this as an initial patch so that folks could review and discuss it.)

I was then able to implement the logic by combining your linked initial implementation with the code linked by @sanket1729 (I particularly like his separation into two functions).

I've intentionally not posted too much code so you can have fun playing with it.

**Comment by Kixunil** [comment]
Perhaps it should return iterator instead? If `Vec` is really required, then `Cow<'u, [T]>` would be a reasonable compromise IMO.

@violet360 please prefer to use iterators, something like this:

```rust
self.unsigned_tx.input.iter().zip(&self.inputs).map(|tx_input, psbt_input| {
// Your code here
})
// collect only if Vec should really be returned (IMO it shouldn't)
.collect::<Result<Vec<_>, _>>()
```

Note this code seems to have significant similarities with [one in my bip78 crate](https://github.com/Kixunil/payjoin/blob/master/bip78/src/psbt.rs#L53) I think we should really try to migrate it here.

**Comment by violet360** [comment]
> So this is a bit of a curly one because of Rust borrowing rules. I played around with this and was able to return a `Result<Prevouts<&TxOut>, Error>` by first changing the `Prevouts::All` variant to use vec instead of a slice (removes the explicit lifetime also).
> 
> ```
> -    All(&'u [T]),
> +    All(Vec<T>),
> ```
> 
> I was unable to work out _why_ we use a slice currently or what the implications of changing this were? (If you choose to follow this route you could do this as an initial patch so that folks could review and discuss it.)
> 
> I was then able to implement the logic by combining your linked initial implementation with the code linked by @sanket1729 (I particularly like his separation into two functions).
> 
> I've intentionally not posted too much code so you can have fun playing with it.

Hi!
I too tried this but again since this one required me to change the existing code base which must have been written taking some future implications into account, hence I did not travel that direction.....

**Comment by violet360** [comment]
> Perhaps it should return iterator instead? If `Vec` is really required, then `Cow<'u, [T]>` would be a reasonable compromise IMO.
> 
> @violet360 please prefer to use iterators, something like this:
> 
> ```rust
> self.unsigned_tx.input.iter().zip(&self.inputs).map(|tx_input, psbt_input| {
> // Your code here
> })
> // collect only if Vec should really be returned (IMO it shouldn't)
> .collect::<Result<Vec<_>, _>>()
> ```
> 
> Note this code seems to have significant similarities with [one in my bip78 crate](https://github.com/Kixunil/payjoin/blob/master/bip78/src/psbt.rs#L53) I think we should really try to migrate it here.

yes... I don't really understand iterators that much, last time used them in c++, I guess RUST has it's own ways maybe...let me learn and see how can I fit this model in here.

**Comment by violet360** [comment]
I thought generating a PR would make easier to review my code .... so did that.

**Comment by Kixunil** [comment]
Yeah iterators in Rust are different (better) and very important because of performance and correctness implications. Definitely try to learn and use them.

You can simply return `impl Iterator<Item=Result<&TxOut, Error>>` and let the caller handle it with `collect()` or such.

**Comment by violet360** [comment]
yeah, ok, thank you, before I start preparing my next commit I have just one question.
If the return type becomes `impl Iterator<Item=Result<&TxOut, Error>>` then what about `sighash::Prevouts` return type as mentioned in the issue, should I just replace that with `impl Iterator<Item=Result<&TxOut, Error>>` ?

**Comment by Kixunil** [comment]
I think it's OK to not use it since the caller can easily create it if actually required. E.g. it's not required in my crate. @sanket1729 feel free to dispute.

**Comment by sanket1729** [comment]
I think it's okay to create an iterator. My crate(the caller) can create
the variant by collecting it.

**Comment by DanGould** [comment]
Was this issue closed by #853 @sanket1729?

Also, wouldn't bip 370 PSBTv2 make this obsolete?

**Comment by Kixunil** [comment]
@DanGould yeah, it was.

PSBTv2 doesn't make it obsolete but our new Rut API should lead to removal of the panic.
